### DISCUSSION TITLE
How can I modify the `rxode2` model structure at a lower level than the piping interface?

### DISCUSSION THREAD

**Discussion by billdenney** [nlmixr2/rxode2#271] [Q&A] [answered]
For the common case of changing the model structure one line at a time or adding/modifying the initial estimates, I know how to do that with `fit %>% model(ka <- exp(lka)) %>% ini(lka <- 0.1)`.

I want to dive deeper into the model.  Specifically, for `nlmixr2lib`, I want to modify the model by looking at it more like a parsed function object where I can go in and make changes.  I found that there is a `fit$fun`, but when I modify it, the changes don't seem to be made.  (My guess is that `fit$fun` is a read-only part of the object.)

How can I make such modifications?

**Comment by mattfidler** [comment, marked as answer]
`fit$fun` is a calculated object, so in essence it is read-only.

However, in theory if you wanted to, you could

1. Assign the function to a new variable  
2. Modify the function however you want .
3. You could evaluate the new function to create a new ui object.


This is what is done in a similar function  in `nlmixr2est`.  Thisuses a `ui$iniDf` and a quoted model block to produce a new ui object.

This is an example of how it works:

https://github.com/nlmixr2/nlmixr2extra/blob/cb1c507f217bd208e003cfe1a461c6bb69bb027b/R/precondition.R#L73-L75

And this is the code:

https://github.com/nlmixr2/nlmixr2extra/blob/cb1c507f217bd208e003cfe1a461c6bb69bb027b/R/precondition.R#L6-L19

**Reply by billdenney** [reply]
Thanks.  Working from the function object itself will work well.

**Comment by dzianismr** [comment]
Hi Matt and Bill,

I am very excited to get started with the newly released rxode2 and nlmixr2. I am particularly interested in testing model modification via piping functionality. Do I understand correctly that rxode2 objects can now be modified via piping as well as nlmixr2 objects (available in the previous releases)? What is the right syntaxis?

Lets take a simple 2cmt model (specified below) as an example. Lets say I would like to add inter individual variability to CL parameter (CL = tvCL*exp(eta.CL)). What is the right way of doing this?

I tried first:
rxRename( ode2cmp, CL = tvCL )
this generates Error: '.data' needs to be a rxUi model

so tried to use functionality of addding a line %>% model(lines, append=TRUE):
ode2cmpt %>% model( lines = "CL = tvCL ", append = TRUE )
and got Error in model.default(., lines = "CL = tvCL ", append = TRUE) : rxode2 does not know how to handle this model statement

Will be really nice to have an example with rxode2 model modification via piping.

ode2cmp <- rxode2({
#additional variables
C2 = A2/V2
C3 = A3/V3

#model
d/dt(A2) =  - CL*C2 - Q*C2 + Q*C3
d/dt(A3)  =            Q*C2 - Q*C3
})

**Reply by mattfidler** [reply]
I think the key here is that it can only modify the `nlmixr2` style UI functions.  So:

```r
ode2cmp <- function(){
  model({
                                        #additional variables
    C2 = A2/V2
    C3 = A3/V3

                                        #model
    d/dt(A2) = - CLC2 - QC2 + QC3
    d/dt(A3) = QC2 - Q*C3
  })
}


ode2cmp %>% rxRename(tvCL=CLC2)

# append
ode2cmp %>% model(CL = tvCL, append = TRUE)

#prepend
ode2cmp %>% model(CL = tvCL, append = NA)
```

Should work  in the development version of `rxode2`  If you want to make it work in the current version of `rxode2` I think you have to add an `ini()` block too.

**Reply by mattfidler** [reply]
I submitted a feature request to allow a simple `rxode2` model to be changed to a `rxUi` object and then it could be piped.

https://github.com/nlmixr2/rxode2/issues/286

**Reply by dzianismr** [reply]
Thanks Matt for the suggested work around and for submitting a feature request.

**Reply by mattfidler** [reply]
This will be available in `rxode2` `2.0.8`

**Reply by mattfidler** [reply]
Which was released to CRAN today.

**Reply by mattfidler** [reply]
... But CRAN will be rolling it back.

**Reply by dzianismr** [reply]
That was implemented at a lightning speed!

**Comment by dzianismr** [comment]
@mattfidler Hi Matt, 

Let me follow up on rxode2 model modification via piping. I have the latest rxode2 2.0.11 from CRAN and would like to make a change in model equations. Lets take a simple 1cmt model example. My goal is to change definition of cl parameter.

ocmt_rx0 <- rxode2( {
     d/dt(depot) = -ka * depot
     d/dt(center) = ka * depot - cl / v * center
     cp = center / v
})

ocmt_rx1 <- rxode2( {
     cl = tvcl
     d/dt(depot) = -ka * depot
     d/dt(center) = ka * depot - cl / v * center
     cp = center / v
})

ocmt_rx2 <- rxode2( {
     d/dt(depot) = -ka * depot
     d/dt(center) = ka * depot - cl / v * center
     cp = center / v
     cl = tvcl
})
ocmt_rx2 will not run, due to the wrong order of the equations (Error in rxSolveSEXP...  : 
  The following parameter(s) are required for solving: cl ).

Dosing and parameters:
et0     <- et(amt=1, time=0 ) %>%
              et(0, 24, by=0.1 ) 
theta0 <- c(ka=0.3, cl = 1, v = 1 )
theta1 <- c(ka=0.3, tvcl = 1, v = 1 )


ocmt_rx0 %>% model( cl <- tvcl*2, append = TRUE ) %>%
   rxSolve( params = theta1, events = et0 )
Will fail, as the code creates code as in ocmt_rx2 with the wrong order of the equations.

ocmt_rx1 %>% model( cl <- tvcl*2 ) %>%
   rxSolve( params = theta1, events = et0 )
Works fine as  correct order of the equations was predefined. 

Is it possible to specify a position in the model code where equation is inserted via model piping?

**Reply by mattfidler** [reply]
Currently this is not possible, but perhaps in the future.  Something like `model(..., append=lhs_variable)` and it will append after the first occurrence of `lhs` variabe

**Reply by dzianismr** [reply]
Thanks, Matt. This would be a very helpful extension.

**Reply by mattfidler** [reply]
This works in the development `rxode2`:

```r
> devtools::load_all()
ℹ Loading rxode2
rxode2 2.0.11.9000 using 4 threads (see ?getRxThreads)
>   ocmt_rx0 <- rxode2( {
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
  })

. + > 
> 
>   m1 <- ocmt_rx0 %>% model( cl <- tvcl*2, append = NA)
ℹ parameter labels from comments are typically ignored in non-interactive mode
ℹ Need to run with the source intact to parse comments
ℹ promote `tvcl` to population parameter with initial estimate 1
> m1
 ── rxode2-based free-form 2-cmt ODE model ─────────────────────────────────────────────────────────────────────── 
 ── Initalization: ──  
Fixed Effects ($theta): 
tvcl 
   1 

States ($state or $stateDf): 
  Compartment Number Compartment Name
1                  1            depot
2                  2           center
 ── Model (Normalized Syntax): ── 
function() {
    ini({
        tvcl <- 1
    })
    model({
        cl <- tvcl * 2
        d/dt(depot) = -ka * depot
        d/dt(center) = ka * depot - cl/v * center
        cp = center/v
    })
}
>   m2 <- ocmt_rx0 %>% model( cl <- tvcl*2, append = d/dt(depot))
ℹ parameter labels from comments are typically ignored in non-interactive mode
ℹ Need to run with the source intact to parse comments
ℹ promote `tvcl` to population parameter with initial estimate 1
> m2
 ── rxode2-based free-form 2-cmt ODE model ─────────────────────────────────────────────────────────────────────── 
 ── Initalization: ──  
Fixed Effects ($theta): 
tvcl 
   1 

States ($state or $stateDf): 
  Compartment Number Compartment Name
1                  1            depot
2                  2           center
 ── Model (Normalized Syntax): ── 
function() {
    ini({
        tvcl <- 1
    })
    model({
        d/dt(depot) = -ka * depot
        cl <- tvcl * 2
        d/dt(center) = ka * depot - cl/v * center
        cp = center/v
    })
}
> 
```

**Reply by mattfidler** [reply]
now you can place lines wherever you want.

**Reply by billdenney** [reply]
@mattfidler Thanks!
